### PR TITLE
Moving ScanIterator to batch/common.

### PR DESCRIPTION
**Pull request by sduskis**
This is part of an effort to move all of the Cloud Bigtable operations into batch/common.  This refactor is part of an effort to create a Beam connector.

### CODE REVIEW

File: bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/batch/common/SplitTracker.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+import com.google.bigtable.repackaged.com.google.protobuf.ByteString;
+
+/**
+ * Determines whether or not a split occurred before a rowKey.
+ *
+ */
+public interface SplitTracker {
+  public boolean isAfterSplit(ByteString rowKey);
+}
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/batch/common/SplitTracker.java]: Will we have multiple implementations of this? If not, I'm not sure the abstraction is worth it

File: bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/batch/common/ScanIteratorFactory.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        }
+        results.add(FLAT_ROW_ADAPTER.adaptResponse(row));
+      }
+      return results.toArray(new Result[results.size()]);
+    }
+
+
+    @Override
+    public boolean isCompletionMarker(Result[] result) {
+      return result == null || result.length == 0;
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/batch/common/ScanIteratorFactory.java]: When will result be null? (same for below)

File: bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/batch/common/SplitTracker.java (hunk 1 above)
Comment (sduskis) [inline bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/batch/common/SplitTracker.java]: There will be 2 implementations: one for beam, one for Dataflow.

File: bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/batch/common/ScanIteratorFactory.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      return result == null ? 0 : result.length;
+    }
+  }
+
+  public static ScanIterator<Result> getSingleResultIterator() {
+    return RESULT_ADVANCER;
+  }
+
+  public static ScanIterator<Result[]> getMultiResultIterator(int resultCount) {
+    return new ResultArrayIterator(resultCount);
```
Comment (sduskis) [inline bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/batch/common/ScanIteratorFactory.java]: We'd have to discuss how to do this with a Source / Reader combination.  They work differently than plain DoFns.

File: bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/batch/common/ScanIteratorFactory.java (hunk 2 above)
Comment (sduskis) [inline bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/batch/common/ScanIteratorFactory.java]: scanner.next() can be null if there are no results.

File: bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/batch/common/ScanIterator.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+ *
+ * @param <ResultOutputType> is either a {@link Result} or {@link Result}[];
+ */
+public interface ScanIterator<ResultOutputType> extends Serializable {
+  /**
+   * Get the next unit of work.
+   * @param resultScanner The {@link ResultScanner} on which to operate.
+   * @param splitTracker The {@link SplitTracker} that determines if a split occurred.
+   */
+  ResultOutputType next(ResultScanner<FlatRow> resultScanner, SplitTracker splitTracker)
```
Comment (sduskis) [inline bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/batch/common/ScanIterator.java]: The ScanIterator is "configuration" that's created in the vm that creates the pipeline.  The ResultScanner and SplitTracker is created per bundle.

**Pull request closed without merging**